evm_loader/program/src/account/state.rs

```diff
@@ -112,30 +124,34 @@ impl<'a> StateAccount<'a> {
         self.account
     }
 
-    pub fn from_account(program_id: &Pubkey, account: AccountInfo<'a>) -> Result<Self> {
-        super::validate_tag(program_id, &account, TAG_STATE)?;
-
-        let (offset, len) = {
-            let header = super::header::<Header>(&account);
-            let offset = BUFFER_OFFSET + header.evm_state_len + header.evm_machine_len;
-            (offset, header.data_len)
-        };
+    pub fn from_account(program_id: &Pubkey, account: &AccountInfo<'a>) -> Result<Self> {
+        super::validate_tag(program_id, account, TAG_STATE)?;
 
-        let data = {
-            let account_data = account.try_borrow_data()?;
-            let buffer = &account_data[offset..(offset + len)];
-            bincode::deserialize(buffer)?
+        let header = super::header::<Header>(account);
+        let data_ptr = unsafe {
+            // Data is more-strictly aligned, but it's safe because we previously initiated it at the exact address.
+            #[allow(clippy::cast_ptr_alignment)]
+            account
+                .data
+                .borrow()
+                .as_ptr()
+                .add(header.data_offset)
+                .cast::<Data>()
+                .cast_mut()
         };
 
-        Ok(Self { account, data })
+        Ok(Self {
+            account: account.clone(),
+            data: ManuallyDrop::new(unsafe { Boxx::from_raw_in(data_ptr, acc_allocator()) }),
+        })
     }
 
     pub fn new(
         program_id: &Pubkey,
         info: AccountInfo<'a>,
```

Comment:
By the time this function is called, the heap is already initialized inside this account.
Add additional checks that anything you write to the account here does not overlap with the heap.

Reply:
I am worrying if writing to the header could overlap the heap.